gh-124165: Optimize keep_top_bit() in heapq module by using _Py_bit_length()

visitorckw opened the pull request:
Optimize the keep_top_bit() function to use _Py_bit_length() instead of manually shifting bits. This allows for more efficient execution on hardware that supports specialized instructions such as bsr on x86. Additionally, the function is now marked as inline, as it is small and only used in one location, improving the potential for inlining by the compiler.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-124165 -->
* Issue: gh-124165
<!-- /gh-issue-number -->

visitorckw committed:
gh-124165: Optimize keep_top_bit() in heapq module by using _Py_bit_length()

Optimize the keep_top_bit() function to use _Py_bit_length() instead
of manually shifting bits. This allows for more efficient execution on
hardware that supports specialized instructions such as bsr on x86.
Additionally, the function is now marked as inline, as it is small and
only used in one location, improving the potential for inlining by the
compiler.

vstinner commented on Modules/_heapqmodule.c:
```diff
@@ -279,16 +280,10 @@ _heapq_heappushpop_impl(PyObject *module, PyObject *heap, PyObject *item)
-        i++;
-    }
-    return n << i;
+    return (Py_ssize_t)1 << (_Py_bit_length((unsigned long)n) - 1);
```
`_Py_bit_length(unsigned long x)` takes an unsigned long, whereas you have a `Py_ssize_t`. On Windows, `long` is only 32-bit, whereas size_t is 64-bit. Your cast is unsafe.

If you want to use _Py_bit_length(), you should add a 64-bit variant using `__builtin_clzll()` and `_BitScanReverse64()`.

visitorckw committed:
gh-124165: Add _Py_bit_length64() for 64-bit integer support

Introduce a new function _Py_bit_length64() to handle 64-bit integers
efficiently. This function utilizes __builtin_clzll() for GCC/Clang and
_BitScanReverse64() for MSVC, ensuring compatibility across different
platforms. For compilers that do not support these intrinsics, a
fallback implementation using a lookup table is provided.

This addition is necessary for safely computing the bit length of
64-bit integers, which is particularly important when working with
types like Py_ssize_t on platforms where size_t is 64-bit, such as
Windows.

vstinner commented on Include/internal/pycore_bitutils.h:
```diff
@@ -180,6 +180,44 @@ _Py_bit_length(unsigned long x)
+    // _BitScanReverse64() is documented to search 64 bits.
+    Py_BUILD_ASSERT(sizeof(uint64_t) <= 8);
+    unsigned long msb;
+    if (_BitScanReverse64(&msb, x)) {
```
There are many warnings:
'_BitScanReverse64' undefined; assuming extern returning int [D:\a\cpython\cpython\PCbuild\_freeze_module.vcxproj]

vstinner commented on Include/internal/pycore_bitutils.h:
```diff
@@ -180,6 +180,44 @@ _Py_bit_length(unsigned long x)
+    else {
+        return 0;
+    }
+#elif defined(_MSC_VER)
```
_BitScanReverse64() is not available on 32-bit Windows (x86, ARM). You should check if you're running 64-bit Windows.

vstinner commented on Include/internal/pycore_bitutils.h:
```diff
@@ -180,6 +180,44 @@ _Py_bit_length(unsigned long x)
+    }
+#elif defined(_MSC_VER) && defined(_WIN64)
+    // _BitScanReverse64() is documented to search 64 bits.
+    Py_BUILD_ASSERT(sizeof(uint64_t) <= 8);
```
You can remove this assertion, I don't see its purpose.

vstinner commented on Include/internal/pycore_bitutils.h:
```diff
@@ -180,6 +180,44 @@ _Py_bit_length(unsigned long x)
+    msb += BIT_LENGTH_TABLE[x];
+    return msb;
+#endif
+}
```
Can you please add tests for this function in Modules/_testinternalcapi.c? Copy check_bit_length() and test_bit_length() to make 64-bit variant and add your test_bit_length64() to module_functions.

vstinner commented on Modules/_heapqmodule.c:
```diff
@@ -279,16 +280,10 @@ _heapq_heappushpop_impl(PyObject *module, PyObject *heap, PyObject *item)
-static Py_ssize_t
+static inline Py_ssize_t
 keep_top_bit(Py_ssize_t n)
 {
```
I suggest to add:

```
Py_BUILD_ASSERT(sizeof(Py_ssize_t) <= sizeof(uint64_t));
```

vstinner commented on Modules/_heapqmodule.c:
```diff
@@ -279,16 +280,10 @@ _heapq_heappushpop_impl(PyObject *module, PyObject *heap, PyObject *item)
-        i++;
-    }
-    return n << i;
+    return (Py_ssize_t)1 << (_Py_bit_length64((uint64_t)n) - 1);
```
You may use _Py_bit_length() on 32-bit platform: when SIZEOF_SIZE_T is 4.

visitorckw committed:
gh-124165: Add test for _Py_bit_length64()

Introduce a new test function, test_bit_length64(), which verifies the
correctness of _Py_bit_length64().
